storage-proofs/porep/src/nse/vanilla/labels.rs

```diff
@@ -486,17 +495,144 @@ fn lc_tree_from_slice<Tree: MerkleTreeTrait>(
 fn tree_from_slice<Tree: MerkleTreeTrait>(
     data: &[u8],
     config: StoreConfig,
-) -> Result<MerkleTree<Tree>> {
-    let mut tree = MerkleTreeWrapper::from_par_iter_with_config(
+) -> Result<LCMerkleTree<Tree>> {
+    let mut tree: MerkleTree<Tree> = MerkleTreeWrapper::from_par_iter_with_config(
         data.par_chunks(NODE_SIZE)
             .map(|node| <Tree::Hasher as Hasher>::Domain::try_from_bytes(node).unwrap()),
         config.clone(),
     )?;
+    let tree_len = tree.len();
+    let leafs = get_merkle_tree_leafs(tree_len, Tree::Arity::to_usize())?;
 
-    // compact the thing
-    tree.compact(config, StoreConfigDataVersion::One as u32)?;
+    // 'v1' compact the existing tree store
+    tree.compact(config.clone(), StoreConfigDataVersion::One as u32)?;
 
-    Ok(tree)
+    // Re-instantiate the 'v1' compacted store as an lc tree.
+    let store = LCStore::new_from_disk(tree_len, Tree::Arity::to_usize(), &config)?;
+    MerkleTreeWrapper::from_data_store(store, leafs)
+}
+
+fn to_gpu_config(conf: &Config) -> gpu::Config {
+    gpu::Config {
+        num_nodes_window: conf.num_nodes_window,
+        num_butterfly_layers: conf.num_butterfly_layers,
+        num_expander_layers: conf.num_expander_layers,
+        degree_expander: conf.degree_expander,
+        degree_butterfly: conf.degree_butterfly,
+        k: conf.k,
+    }
+}
+
+type GPUHasher = storage_proofs_core::hasher::PoseidonHasher;
+type GPUTree = storage_proofs_core::merkle::OctLCMerkleTree<GPUHasher>;
+pub fn encode_with_trees_all<'a, Tree: 'static + MerkleTreeTrait, I>(
+    conf: &Config,
+    rows_to_discard: usize,
+    inps: I,
+) -> Result<Vec<(Vec<LCMerkleTree<Tree>>, LCMerkleTree<Tree>)>>
+where
+    I: Iterator<
+        Item = (
+            Vec<StoreConfig>,
+            u32,
+            &'a <Tree::Hasher as Hasher>::Domain,
+            &'a mut [u8],
+        ),
+    >,
+{
+    let gpu_conf = to_gpu_config(conf);
+    let pool = gpu::utils::all_devices()
+        .map_err(|e| -> gpu::NSEError { e.into() })
+        .and_then(|devs| {
+            Ok(gpu::SealerPool::new(
+                devs,
+                gpu_conf,
+                gpu::TreeOptions::Enabled { rows_to_discard },
+            )?)
+        });
+
+    if !settings::SETTINGS.lock().unwrap().use_gpu_nse
```

Comment:
Instead of checking the setting value here, we should pass it as an input to `encode_with_trees_all`. We might need another wrapper function just for that purpose. This will make it easier to test both cases for equivalence.